### ISSUE TITLE
Plugin doesn't scale nodes

### ISSUE THREAD

**Issue by kkurzyna** [jenkinsci/ec2-fleet-plugin#43]
Hi,

I have noticed, that scaling isn't working correctly. 

Case 1: There are few builds in queue
Plugin: Doesn't create new nodes and doesn't send to AWS request to increase fleet capacity (even thought spawning them wouldn't exceed max-cluster-size);

Case 2: Min cluster size is greater than 1
Plugin: Again, it will only create 1 node and will not spawn more.

Case 3: AWS spot fleet size is 2 and min cluster size is 2
Plugin: After some time Plugin will set AWS spot fleet capacity to 1, but will keep min cluster size at 2

The weirdest part is, those are some recent bugs - I have used this plugin for some time and I haven't notice such an issues earlier.

If someone knows, what is going on - answer please (or just bugifx it, if it really is a problem with plugin).

**Comment by daniilyar** [comment]
All the above reproduces for us also. Didn't find the line of code that actually scales fleet up from 0 to 1 so probably it is simply not implemented..

**Comment by michaljrk** [comment]
I can confirm the same happens for me, (I queued 5-6 execution of the same job, assigned to the fleet agents only and no scale out happened, the jobs just remained waiting). 

It would much appreciate to get any update on that issue.

**Comment by akennealy** [comment]
I can also reproduce it... builds in Queue but no fleet nodes spinning up to take on the job. 

Jenkins 150.1
eC2 fleet plugin 1.1.8

**Comment by vliubko** [comment]
Same situation.

Logs shows nothing, when I change minimal number of nodes
Also, when I manually set capacity of a Fleet to 0, plugin didn't do anything.

Only repeatable status check without any action
```
Jan 15, 2019 9:37:53 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
Fleet Update Status called
Jan 15, 2019 9:37:53 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
# of nodes:1
Jan 15, 2019 9:37:53 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
Fleet (ec2-fleet) contains instances [i-052bc465b309c8c8a]
Jan 15, 2019 9:37:53 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
Jenkins contains dying instances []
Jan 15, 2019 9:38:03 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
Fleet Update Status called
Jan 15, 2019 9:38:03 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
# of nodes:1
Jan 15, 2019 9:38:03 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
Fleet (ec2-fleet) contains instances [i-052bc465b309c8c8a]
Jan 15, 2019 9:38:03 AM FINE com.amazon.jenkins.ec2fleet.EC2FleetCloud
Jenkins contains dying instances []
```

**Comment by andrea-vaccaro** [comment]
Hi,
we have replicated the problem with Jenkins 2.150.1 and EC2 Fleet plug-in 1.1.9: after the fleet is scaled down because of idle instances, new instances are not added to the fleet when needed, even though the number of existing instances is less then max cluster size.

@Plugin developers
Have you analyzed the problem?
Do you know if it possible to use a workaround?

We could disable the scaling mechanism but keeping the spot instances always running will cost too much.

Thanks,
Andrea

**Comment by bertjwregeer** [comment]
I am running the latest version of the plugin, with min cluster size: 1, and max cluster size: 5.

And scaling up/down works without issues.

**Comment by akennealy** [comment]
Same here, but this is a workaround for me -- I want to be able to set min to 0 so I'm not paying for 24/7 (spot) instances. I have to set it to 1 or else it wont scale up as one job comes into the queue.

**Comment by YOO629** [comment]
jenkins v2.138.2, ec2-spot-fleet-plugin v1.1.9, no issues here. scales up or down as expected. I haven't tried setting min cluster size to 0 since I want to keep a certain number of instances ready to go (scaling up tends to take quite a bit of time)

**Comment by chopmann** [comment]
I added Full-ec2 to the instance running the jenkins-master and created a FleetRequest with '0' as Target through cloudformation. Don't know which of the things solved my Issue (I think giving the full-ec2 to the master did it). Now plugin connects and behaves as expected.

**Comment by ianfixes** [comment]
Should I add any of this info to my documentation patch? https://github.com/jenkinsci/ec2-fleet-plugin/pull/56

**Comment by ebalakumar** [comment]
> jenkins v2.138.2, ec2-spot-fleet-plugin v1.1.9, no issues here. scales up or down as expected. I haven't tried setting min cluster size to 0 since I want to keep a certain number of instances ready to go (scaling up tends to take quite a bit of time)

I am using the same version . Setting min  cluster size to 0 is not downsizzing the instance

**Comment by rnhurt** [comment]
I'm experiencing this same problem.  My spot slave scales up/down just fine when the MIN instance is > 0.  Setting it to 0 causes the spot slave to terminate (after the idle timeout).  

When I run another job with that label, the job pauses (like it should) and reports that there are no instances with that label (like it should).  I've waited for 2 hours for the spot slave to startup but it never did.  Once I updated the MIN instance to 1 the spot slave started and the job ran.

**Comment by chopmann** [comment]
Probably to "wide" but this got my setup working:
![jenkins_conf](https://user-images.githubusercontent.com/2122863/56347252-2bedcd80-61c4-11e9-81ae-2305a87f0401.png)


```yaml
Resources:
  Ec2Role: # For Jenkins-Master
    Type: AWS::IAM::Role
    Properties:
      RoleName: !Sub 'Ec2Role-${AWS::StackName}'
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
        - Effect: Allow
          Principal:
            Service:
            - 'ec2.amazonaws.com'
          Action:
          - 'sts:AssumeRole'
      Path: '/'
      Policies:
        - PolicyName: SpotAgentPolicy
          # Policy is documented at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet-requests.html#spot-fleet-iam-users
          PolicyDocument:
            Version: 2012-10-17
            Statement:
            - Effect: Allow
              Action:
              - ec2:*
              - iam:ListInstanceProfiles
              - iam:ListRoles
              - iam:PassRole
              Resource: "*"

  IAMUserJenkins:
    Type: AWS::IAM::User
    Properties:
      UserName: XXXXX #  jenkins-fleet-user
      Policies:
      - PolicyName: SpotAgentPolicy
        # Policy is documented at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet-requests.html#spot-fleet-iam-users
        PolicyDocument:
          Version: 2012-10-17
          Statement:
          - Effect: Allow
            Action:
            - ec2:*
            - iam:ListInstanceProfiles
            - iam:ListRoles
            - iam:PassRole
            Resource: "*"

  LaunchTemplate:
    Type: "AWS::EC2::LaunchTemplate"
    Properties:
      LaunchTemplateName: "jenkins-slave-small"
      LaunchTemplateData:
        # CreditSpecification: # needed for t2-t3
        #   CpuCredits: standard
        ImageId: "ami-XXXXXXX" # ci-jenkins-slave-base (ubuntu)
        InstanceType: "m3.medium"
        InstanceMarketOptions:
          SpotOptions:
            MaxPrice: "0.015"
        KeyName: !Ref KeyName
        IamInstanceProfile:
          Arn: !ImportValue 'jenkins-security-SlaveProfile'
        NetworkInterfaces:
        - DeviceIndex: 0
          AssociatePublicIpAddress: true
          Groups:
            - !ImportValue 'jenkins-security-SlaveSG'
          DeleteOnTermination: true

  spotFleet:
    DependsOn:
      - LaunchTemplate
    Type: AWS::EC2::SpotFleet
    Properties:
      SpotFleetRequestConfigData:
        ExcessCapacityTerminationPolicy: Default
        AllocationStrategy: lowestPrice
        IamFleetRole: !ImportValue 'jenkins-security-SlaveRole'
        LaunchTemplateConfigs:
          - LaunchTemplateSpecification:
              LaunchTemplateId: !Ref LaunchTemplate
              Version:
                Fn::GetAtt:
                  [ "LaunchTemplate", "LatestVersionNumber" ]
            Overrides:
              - SubnetId: 
                  Fn::ImportValue: !Sub '${ParentVpcStack}-PublicAId'
          - LaunchTemplateSpecification:
              LaunchTemplateId: !Ref LaunchTemplate
              Version:
                Fn::GetAtt:
                  [ "LaunchTemplate", "LatestVersionNumber" ]
            Overrides:
              - SubnetId: 
                  Fn::ImportValue: !Sub '${ParentVpcStack}-PublicBId'
        TargetCapacity: 0
        TerminateInstancesWithExpiration: true
```

**Comment by rnhurt** [comment]
@chopmann, I have a couple of questions about your configuration:
* does it allow autoscaling to zero (0) instances?  
* does it allow you to run a successful "Test Connection"?

I've updated my IAM permissions to match yours (actually, both the Jenkins master and slave(s) have almost every IAM permission right now) but I still can't get the ASG to go to 0, nor does my "Test Connection" button work properly.  :(

**Comment by chopmann** [comment]
@rnhurt 

1. yes
2. yes

I also deployed the Jenkins-Master on the same vpc:
Based of https://github.com/bwolf/aws-jenkins (@bwolf) with some small mods to work on t3.small and gp2 EBS

**Comment by rnhurt** [comment]
Ah, well this may be the difference.  My Jenkins master and slave nodes are on completely different AWS accounts; the master node is in our "shared" environment and the "slave" node is in our integration environment.  

Still, my permissions seem correct: 
* the plug-in auto-scale the fleet up without any problems
* I can use the same credentials to get a list of spot fleet requests sucessfully from the CLI on my laptop (`aws ec2 describe-spot-fleet-requests`)
* when I remove the access from the credentials, the plugin stops responding completely (ie. won't update the region list) and gives me a "hostname cannot be null" error.
* when I add the access back to the credentials, the plugin allows me to configure it again but the test gives me the same old "SpotFleet Request(s) don't exist..." error.

Here are the recent events from my SpotFleet instance showing that it receives a fleetRequestChange for the scale up but nothing from the last 10 minutes where there are no jobs being run and the executers are idle.

Timestamp  | Event Type | Status | Description 
------------|-------------|-------|---------
4/18/2019, 12:50:16 PM | bidChange | active | BidId sir-ccgr4q9h, PreviousState: active |  
4/18/2019, 12:50:15 PM | instanceChange | launched | {"instanceType":"t3.small","image":"ami-afd15ed0","productDescription":"Linux/UNIX","availabilityZone":"us-east-1c"}
4/18/2019, 12:50:13 PM | bidChange | submitted | BidId sir-ccgr4q9h, PreviousState: open |  
4/18/2019, 12:50:12 PM | fleetRequestChange | modify_succeeded | New targetCapacity: 1
4/18/2019, 12:49:24 PM | fleetRequestChange | modify_in_progress | Modify request received. Requested targetCapacity: 1

**Comment by rnhurt** [comment]
Ok, I think I have a bit more information on the problem now.  It appears as if the plugin doesn't update the SpotFleet configuration properly (or something).  If you change the Minimum Cluster Size while there are slave instances running then it doesn't reflect that change.

This is how I corrected the problem of not scaling down:
1. Change the Minimum Cluster Size to 0 (zero) and save the changes
1. Open the AWS Console and change the SpotFleet request to 0 (zero) manually
1. Once all the slave instances die go back to Jenkins and start a build on the slave node(s)
1. The SpotFleet should update its status to run the job and then scale down to 0 (zero) when the idle timeout expires.

You can also see this effect if you change the Min. Cluster Size from 0 (zero) to something else.  I changed mine from 0 (zero) to 1 and nothing changed; the SpotFleet config didn't change and the "EC2 Fleet Status" in Jenkins still listed the "target" nodes as 0.  However, running a build on the slave node caused everything to update to 1, as expected.

I think the problem is that changing the Min. Cluster Size is not triggering a change on the AWS SpotFleet configuration.  It's not until a change is forced (through manually killing things or causing a job to build) that the SpotFleet config gets updated.

**Comment by Puneeth-n** [comment]
@rnhurt Thanks for the investigation and I can confirm that it works. I also enabled FINE logging for the plugin and could see that it terminated and set the fleet to 0 after the idle termination timeout of 15 minutes that I had set

**Comment by Bryksin** [comment]
Have the same problem!

1. I configured the fleet to star from 1 and scale to 5 with 2 build slots per node
2. node was found and registered successfully
3. created dummy job to sleep for 5 min
4. executed 5 builds in Q
5. slaves didn't scale, target count on the main page also didn't change, in AWS Console fleet request also didn't change

So autoscaling based on Q didn't scale
Changing `Minimum Cluster Size` to any number bigger than existing 1 also didn't help. The target remains as 1

I was following the @rnhurt suggestion, and it did work to some degree, basically, it started spawning nodes, but it was not attempting to access it and install agents, it spawn 5 nodes and kept them offline, I had to click on each one and select install agent button, only after that they become online (maybe it is another bug based on enabled: `Only running instances` - `Experimental Add EC2 instance to slaves only when state is running`)

**Comment by Bryksin** [comment]
Just to add, yesterday evening I left 1 node running on purpose where the target was set to zero and 30 min delay before scaledown, in the morning I discovered that scaledown didn't happen, the instance still up and running!

basically, it is sad to admit that this plugin completely uselessly, it doesn't scale nor up nor down :(

**Comment by terma** [comment]
hi @Bryksin

Could you please check conditions to be able downscale:
- plugin configure with ```idleMinutes > 0``` 
- and the required time is passed and the node didn't run any build during this time, you can check this if you open node page and goto ```builds``` tab

**Comment by Bryksin** [comment]
Hi @terma 

as I mentioned above idleMinutes were set to 30, jenkins was left alone overnight with no running jobs, means it was 100% idle, in the morning agent was not terminated (not scaled down) 

Btw you linked my another ticket #106 to this issue, but had to #32, not #43

**Comment by terma** [comment]
thx for correction, can I ask what version of jenkins do you use?

**Comment by Bryksin** [comment]
it is default docker LTS image `jenkins/jenkins:lts` nothing customised, taken as it is

Run with this command:

```
sudo docker run -d --rm --name jenkins-master -v /var/jenkins_home:/var/jenkins_home -v /var/run/docker.sock:/var/run/docker.sock -p 80:8080 -p 5000:5000 jenkins/jenkins:lts
```

**Comment by terma** [comment]
hi team,

I think I know root cause of all of this problems with scale up/down, zero smt. Please find explanation and manual fix below.

I'm going to close this issue and create new for **long term fix #111**. Feel free to comment this one if you see problem with scale down/up and it doesn't feed case below.

### In short

Changed plugin settings will not be applied to existent nodes, you need to restart Jenkins or delete nodes, so plugin will bring them back.

### Long

#### Problem

Jenkins has special lifecycle for plugins like EC2 Fleet which in fact implementing ```hudson.slaves.Cloud```. Any time when settings need to be updated, Jenkins create new instance of plugin and forget about old, in other word plugin settings are immutable. 

Part which responsible for scale down  ```com.amazon.jenkins.ec2fleet.IdleRetentionStrategy``` get link to plugin instance (settings) when created, and only once. As result, new settings will be applied only to new nodes, as new nodes will be created with link to new plugin instance, while existent nodes still have link to old plugin instance (settings).

#### Temporary fix

- Restart Jenkins
- **or** manually delete nodes in Jenkins UI by clicking on them and select delete
  - no worries it will delete only Jenkins representation of EC2 instance, not actual one, and plugin will recreated them in ```10 sec``` that's refresh time

#### Real fix

There are couple of options, to transfer retention policies to new instance of plugin, required a little bit more investigation.

**Comment by fruch** [comment]
so in other words, you need to create the configuration correct for the first time, and it would work as expected ?

**Comment by terma** [comment]
Yes. As soon as the configuration is correct it should work as expected. You have a few options for configuration: create it, or update and manually delete all slaves from Jenkins, or update and restart Jenkins

**Comment by ryanaslett** [comment]
I believe that there are some other issues as well where if you change your plugin settings, any existing running nodes lose their config or they get reset to some default values.

I've had this plugin successfully processing builds and starting/stopping nodes for more than a week, with a 3 minute timeout.

Today I added another cloud configuration and another spot fleet request to go with it for a different type of job, and the *old* nodes associated with the working config are no longer timing out. I have some that have been sitting for 2 hours without processing a build.

So yes, long story short, you cannot/should not manipulate the config of this plugin if you have any running instances.

**Comment by fruch** [comment]
As a work around solution (at least to my case that I'm using those for parallel pipeline jobs, i.e. splitting tests), I change the target capacity from within my pipeline job, and then the downsize seems work as expected.

**Comment by terma** [comment]
hi all,

Fix for incorrect settings after their modification released under version [1.9.2](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.9.2)

Fixed version will apply new settings imediatelly when you pressed ```Save``` configuration button. Regarless if you have running slaves or not.

Please ping me/create issue if you still see this problem.

**Comment by sdhoka** [comment]
@terma  
I am still running into this issue with version 1.9.2. When I tried to save my plugin config, I found the following error in my fleet plugin log
`Failed to save /var/lib/jenkins/hudson.plugins.groovy.Groovy.xml
java.lang.UnsupportedOperationException: Refusing to marshal java.util.concurrent.atomic.AtomicInteger for security reasons; see https://jenkins.io/redirect/class-filter/
	at hudson.util.XStream2$BlacklistedTypesConverter.marshal(XStream2.java:543)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
	at hudson.util.RobustReflectionConverter.marshallField(RobustReflectionConverter.java:265)
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:252)
Caused: java.lang.RuntimeException: Failed to serialize hudson.plugins.groovy.AbstractGroovy$AbstractGroovyDescriptor#instanceCounter for class hudson.plugins.groovy.Groovy$DescriptorImpl
	at hudson.util.RobustReflectionConverter$2.writeField(RobustReflectionConverter.java:256)
	at hudson.util.RobustReflectionConverter$2.visit(RobustReflectionConverter.java:224)
	at com.thoughtworks.xstream.converters.reflection.PureJavaReflectionProvider.visitSerializableFields(PureJavaReflectionProvider.java:138)
	at hudson.util.RobustReflectionConverter.doMarshal(RobustReflectionConverter.java:209)
	at hudson.util.RobustReflectionConverter.marshal(RobustReflectionConverter.java:150)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:43)
	at com.thoughtworks.xstream.core.TreeMarshaller.start(TreeMarshaller.java:82)
	at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.marshal(AbstractTreeMarshallingStrategy.java:37)
	at com.thoughtworks.xstream.XStream.marshal(XStream.java:1026)
	at com.thoughtworks.xstream.XStream.marshal(XStream.java:1015)
	at com.thoughtworks.xstream.XStream.toXML(XStream.java:988)
	at hudson.XmlFile.write(XmlFile.java:193)
Caused: java.io.IOException
	at hudson.XmlFile.write(XmlFile.java:200)
	at hudson.model.Descriptor.save(Descriptor.java:872)
	at hudson.plugins.groovy.Groovy$DescriptorImpl.configure(Groovy.java:195)
	at jenkins.model.Jenkins.configureDescriptor(Jenkins.java:3731)
	at jenkins.model.Jenkins.doConfigSubmit(Jenkins.java:3695)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:343)
	at org.kohsuke.stapler.interceptor.RequirePOST$Processor.invoke(RequirePOST.java:77)
	at org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:184)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:117)
	at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:129)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:734)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:864)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:668)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:860)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1650)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:154)
	at org.jenkinsci.plugins.ssegateway.Endpoint$SSEListenChannelFilter.doFilter(Endpoint.java:225)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at io.jenkins.blueocean.auth.jwt.impl.JwtAuthenticationFilter.doFilter(JwtAuthenticationFilter.java:61)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at io.jenkins.blueocean.ResourceCacheControl.doFilter(ResourceCacheControl.java:134)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at com.smartcodeltd.jenkinsci.plugin.assetbundler.filters.LessCSS.doFilter(LessCSS.java:47)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at jenkins.metrics.impl.MetricsFilter.doFilter(MetricsFilter.java:125)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:239)
	at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:215)
	at net.bull.javamelody.PluginMonitoringFilter.doFilter(PluginMonitoringFilter.java:88)
	at org.jvnet.hudson.plugins.monitoring.HudsonMonitoringFilter.doFilter(HudsonMonitoringFilter.java:114)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.plugins.greenballs.GreenBallFilter.doFilter(GreenBallFilter.java:59)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:151)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:99)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:190)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:530)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:347)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:256)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
`

**Comment by terma** [comment]
Hi,

I'm wondering how did you send this message, I don't see any thread/issue
on Github about that, but the email said from GitHub =)

Issue intro (probably you already know that): When you press save button on
Jenkins configuration page it actually saves a configuration for all
plugins regardless of actual changes. Jenkins saves the configuration by
serializing plugins objects into XML. As result Jenkins sensitive on what
references do you have in plugin object, for example, if you have
references to some complex object Jenkins serializer must to traverse all
objects and represent them in XML. For some class like Atomic it's very
hard work as it depends on other threads objects. That's why Jenkins has
whitelist of classes which could be stored otherwise it fail.

About the issue: the plugin doesn't use AtomicInteger for production code,
only for tests which are not part of plugin jar for Jenkins. Based on error
message, I assume we have problem with hudson.plugins.groovy.Groovy.xml plugin,
so I suggest you to take a look on your groovy script or ticket
https://wiki.jenkins.io/display/JENKINS/Groovy+plugin another solution just
whitelist this class (link in error should explain how) but I don't suggest
that

Thx,
Artem

**Comment by ZillaG** [comment]
I have jenkns v2.164.3 and Ec2 Fleet jenkins plugin v1.10.0, and my fleet does not scale up. I have min cluster size to 1, and max to 5. In AWS, my fleet capacity is 10.

**Comment by terma** [comment]
@zillag this could be because of many reasons, do y see any logs about your fleet in jenkins log?

**Comment by ricardoribas** [comment]
I don't know if this problems is already solved or not. But i had the same problem on the last few days. But i think the reason is related with the cloudformation template updates. If you update the cloudformation template, the jenkinsSpotFleet ids will change. So, to solve this problem i just updated the ids in the cloud settings and you are good to go again.
